### PR TITLE
Parse incomplete code differently to better reflect user intent.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/15881

This is a port of a parsing strategy we took in TypeScript to better deal with code as the user is typing it.  Note that this is a change in how we parse code in *error*, and it involves heuristics to make the tree better match the user intent.  The code in question is code like:

```c#
Task.
await Task.Delay(...);
```

Today, the C# parser eagerly parses this as a local declaration of the form "Task.await Task".  i.e. the "Type" is "Task.await" and the VariableDeclarator is "Task".  This clearly doesn't match what the user intends, and it messes up higher layers of the stack.

Specifically, because "Task." is a QualifiedName, that changes how we treat it (when it really should be a MemberAccessExpression).  Similarly, because it looks like we're declaring a local called 'Task', we introduce a bogus LocalSymbol into scope, which messes up binding of names like "Task" (it finds the local instead of the type).

The fix is to tweak how we parse here.  We specifically look for the pattern:

```
Id dot new-line
Id id
```

And we do not think of it as a LocalDeclaration *unless* we see a following token that more definitely demonstrates that it is local-variable.  i.e.

```
X.
Y z;
```

In this case, there is no syntax error, so we have to accept this code as being a local variable declaration.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler 

@agocke To make sure i didn't break parsing of LocalFunctions.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler Where is the best place to put tests for this?

**Comment by CyrusNajmabadi** [comment]
I'm going to add tests to ExpressionParsingTests for this.

**Comment by CyrusNajmabadi** [comment]
Note: in these error situations we use the presence of the newline as a strong indicator of user intent.  This works quite well in practice and makes the code parse intuitively as the user thinks it should.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs [hunk 1]
```diff
@@ -2267,5 +2267,544 @@ private bool IsValid_a()
...
+                            N(SyntaxKind.ExpressionStatement);
+                            {
+                                N(SyntaxKind.SimpleMemberAccessExpression);
+                                {
+                                    N(SyntaxKind.IdentifierName);
+                                    {
+                                        N(SyntaxKind.IdentifierToken, "Task");
+                                    }
+                                    N(SyntaxKind.DotToken);
+                                    M(SyntaxKind.IdentifierName);
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ExpressionParsingTests.cs]: note the missing identifier here.  we now parse this as ```Task.<missing>``` as would be expected from the given source.

**Comment by CyrusNajmabadi** [comment]
See https://github.com/Microsoft/TypeScript/blob/master/src/compiler/parser.ts#L1923 as an example of how we dealt with this in TypeScript.

This issue was more pronounced in TS/JS as the following is legal in their language:

```foo.any_keyword```

So we would commonly have the first keyword of the next line getting sucked into the parse of the previous unterminated construct.  This was an elegant way to address the issue low in the stack so that all higher levels worked well.

**Comment by CyrusNajmabadi** [comment]
Tagging @gafter for parser changes.

**Comment by CyrusNajmabadi** [comment]
Ok, tests have been added.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -7010,6 +7010,53 @@ private bool IsPossibleLocalDeclarationStatement(bool allowAnyExpression)
...
+                        //
+                        //      X.
+                        //      Y z
+                        //
+                        // This is only a local declaration if we have:
+                        //
+                        //      X.Y z;
+                        //      X.Y z = ...
+                        //      X.Y z, ...  
+                        //      X.Y z( ...      (local function)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: There is also the case `z[` which is an erroneous local array declaration.

**Comment by gafter** [comment]
@CyrusNajmabadi `master` is in ask mode for RC3. Are you confident this would pass ask mode?

**Comment by CyrusNajmabadi** [comment]
Fairly confident yes. This is a pretty reasonable situation and the experience is currently poor

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: sure!

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Test added.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Can we have a second review here, please?

**Comment by AlekseyTs** [comment]
> It would be nice to see the errors associated with the parse tests. I'll enhance the tests with that info in a separate PR.

Why do we need to do this in a separate PR? I think it is appropriate to do this in this PR, this make it easier to review and sign-off on the issue. @gafter If you want to help @CyrusNajmabadi, you can push commit to this PR.

**Comment by CyrusNajmabadi** [comment]
I would also be fine with more commits pushed to this.

**Comment by gafter** [comment]
@CyrusNajmabadi Don't wait for me. I'm technically on vacation this month, and this is not the next thing I would do. If you and/or @AlekseyTs think that needs to be done as part of this PR (I don't), someone other than me will have to do it.

**Comment by CyrusNajmabadi** [comment]
Oh ok.  I'll do taht and let you know when it's done.

**Comment by CyrusNajmabadi** [comment]
@AlekseyTs Tests pushed.

**Comment by CyrusNajmabadi** [comment]
retest windows_eta_open_prtest  please

**Comment by CyrusNajmabadi** [comment]
Tagging @srivatsn For approval once @AlekseyTs Has taken a look and signed off.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 3]
```diff
@@ -2830,6 +2833,7 @@ class C
             End Using
         End Function
 
+        ' <WpfFact, Trait(Traits.Feature, Traits.Features.Completion)>
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: @rchande I tried to enable this test but it fails.  Can you take a look and fix/enable the test?

**Comment by AlekseyTs** [comment]
LGTM. However, please consider if it is better to let the old remaining logic in IsPossibleLocalDeclarationStatement to be applied to the situation that the new code recognized as a possible valid declaration.

**Comment by CyrusNajmabadi** [comment]
You're totally right aleksey.  I'll make that change and resubmit

**Comment by AlekseyTs** [comment]
LGTM

**Comment by CyrusNajmabadi** [comment]
@gafter Are you ok with me merging this in?  You assigned to yourself, so i wasn't sure if there was anything you wanted to do with this.

**Comment by CyrusNajmabadi** [comment]
Merging in as i see that @gafter LGTMed this.

**Pull request merged**